ownCloudAppShared/Intent/SaveFileIntentHandler.swift

```diff
@@ -0,0 +1,151 @@
+		if AppLockHelper().isPassCodeEnabled {
+			completion(SaveFileIntentResponse(code: .authenticationRequired, userActivity: nil))
+		} else {
+			if let path = intent.path, let uuid = intent.account?.uuid, let file = intent.file, let fileURL = file.fileURL {
+				let accountBookmark = OCBookmarkManager.shared.bookmark(for: uuid)
+
+				if let bookmark = accountBookmark {
+					OCCoreManager.shared.requestCore(for: bookmark, setup: nil, completionHandler: { (core, error) in
+						if error == nil, let core = core {
+							self.itemTracking = core.trackItem(atPath: path, trackingHandler: { (error, item, isInitial) in
+								if let targetItem = item {
+									var newFilename = file.filename
+									if let filename = intent.filename as NSString?, filename.length > 0, let defaultFilename = file.filename as NSString? {
+										var pathExtention = defaultFilename.pathExtension
+										if let fileExtension = intent.fileextension, fileExtension.count > 0 {
+											pathExtention = fileExtension
+										}
+										if let changedFilename = filename.appendingPathExtension(pathExtention) {
+											newFilename = changedFilename
+										}
+									} else if let fileExtension = intent.fileextension, fileExtension.count > 0, let defaultFilename = file.filename as NSString? {
+										let filename = defaultFilename.deletingPathExtension as NSString
+										if let changedFilename = filename.appendingPathExtension(fileExtension) {
+											newFilename = changedFilename
+										}
+									}
+
+									if core.importFileNamed(newFilename,
+															at: targetItem,
+															from: fileURL,
+															isSecurityScoped: true,
+															options: [OCCoreOption.importByCopying : true],
+															placeholderCompletionHandler: { (error, item) in
+																if error != nil {
+																	completion(SaveFileIntentResponse(code: .failure, userActivity: nil))
+																}
+									},
+															resultHandler: { (error, _ core, _ item, _) in
+																if error != nil {
+																	completion(SaveFileIntentResponse(code: .failure, userActivity: nil))
+																} else {
+																	completion(SaveFileIntentResponse.success(filePath: item?.path ?? ""))
+																}
+									}
+										) == nil {
+										completion(SaveFileIntentResponse(code: .failure, userActivity: nil))
+									}
+								} else {
+									completion(SaveFileIntentResponse(code: .pathFailure, userActivity: nil))
+								}
+
+								if isInitial {
+									self.itemTracking = nil
+								}
+							})
+
+						} else {
+							completion(SaveFileIntentResponse(code: .failure, userActivity: nil))
+						}
+					})
+				} else {
+					completion(SaveFileIntentResponse(code: .accountFailure, userActivity: nil))
+				}
+			} else {
+				completion(SaveFileIntentResponse(code: .failure, userActivity: nil))
```

Comment:
@hosy less nesting possible?

Reply:
Looks strange, but there are different return codes for the completion handlers. I think cannot be reduced, if I want to give the correct failure feedback top the user.